### PR TITLE
network: Fix set bond device MAC address failed

### PR DESCRIPTION
**Pull request by zhangjian3032**
Issue:
When device is in bond mode and booting up, there is a probability of set bond MAC address failed due to `Device or resource busy` error.

In systemd-networkd, set MAC address steps are:
1. Try to set MAC address to device.
2. If failed with `Device or resource busy`, then `Down` the device.
3. Try to set MAC address to device again.

Currently, Even down the bond device, the bond device is still return `Device or resource busy` error. So the MAC address set failed.

The root cause is that this not enough to down the bond device. We need to down all the slaves of the bond device.
About this descprition, we could use those commands to check:
```shell
We have two network devices: eth0, bond1, eth0 is slave of bond1.
They are all up.

1. Down bond1, and set MAC address to bond1.
~# ip link set bond1 down
~# ip link set bond1 address 00:11:22:33:44:55
ip: SIOCSIFHWADDR: Device or resource busy

2. Down eth0, and set MAC address to bond1.
~# ip link set eth0 down
~# ip link set bond1 address 00:11:22:33:44:55
Set okay.
```

Fix:
When setting the mac for the second time, if the device kind is bond, then we need to down the slave devices of bond device.

Tested: Verified in a long time test( reboot cycles ).

Fixes: #25627

Signed-off-by: Jian Zhang <zhangjian.3032@bytedance.com>

### CODE REVIEW

File: src/network/networkd-setlink.c [hunk 1]
```diff
@@ -502,6 +502,14 @@ static int link_is_ready_to_set_link(Link *link, Request *req) {
                         r = link_down_now(link);
                         if (r < 0)
                                 return r;
+
+                        /* If the kind of the link is "bond", we need
+                         * set the slave link down as well. */
+                        if (link->kind && streq_ptr(link->kind, "bond")) {
```
Comment (yuwata) [inline src/network/networkd-setlink.c]: `streq_ptr()` gracefully handles the case that `link->kind` is `NULL`. Hence, please simply do like the following:
```c
if (streq_ptr(link->kind, "bond")) {
```

**Review by yuwata — CHANGES_REQUESTED** [review]
LGTM, except for one minor point.
Could you add a simple test to test/test-network/systemd-networkd-tests.py ?

File: test/test-network/conf/23-bond199.network [hunk 2]
```diff
@@ -4,3 +4,6 @@ Name=bond199
 
 [Network]
 IPv6AcceptRA=no
+
+[Link]
+MACAddress=00:11:22:33:44:55
```
Comment (yuwata) [inline test/test-network/conf/23-bond199.network]: Please append the line break at the end.

File: test/test-network/systemd-networkd-tests.py [hunk 3]
```diff
@@ -3882,6 +3882,15 @@ def test_bond_primary_slave(self):
...
         self.assertRegex(output, 'primary dummy98')
 
+    def test_bond_mac_config(self):
+        copy_network_unit('23-primary-slave.network', '23-bond199.network', '25-bond-active-backup-slave.netdev', '12-dummy.netdev')
+        start_networkd()
+        self.wait_online(['dummy98:enslaved', 'bond199:degraded'])
+
+        output = check_output('ip -d link show bond199')
+        print(output)
+        self.assertRegex(output, 'link/ether 00:11:22:33:44:55')
```
Comment (yuwata) [inline test/test-network/systemd-networkd-tests.py]: Please use `self.assertIn('...', output)`.

File: test/test-network/systemd-networkd-tests.py [hunk 4]
```diff
@@ -3882,6 +3882,15 @@ def test_bond_primary_slave(self):
         print(output)
         self.assertRegex(output, 'primary dummy98')
 
+    def test_bond_mac_config(self):
```
Comment (yuwata) [inline test/test-network/systemd-networkd-tests.py]: Please merge this with the above `test_bond_primary_slave()`.

Also, maybe, to reproduce the reported issue, first configure the bond interface without `MACAddress=` setting, then create an drop-in for `23-bond199.network`, e.g. `23-bond199.network.d/mac.conf` with
```
[Link]
MACAddress=00:11:22:33:44:55
```
then call `networkctl reload`, and check if the MAC address is applied.

File: test/test-network/conf/23-bond199.network (hunk 2 above)
Comment (zhangjian3032) [reply test/test-network/conf/23-bond199.network]: okay, done

File: test/test-network/systemd-networkd-tests.py (hunk 3 above)
Comment (zhangjian3032) [reply test/test-network/systemd-networkd-tests.py]: okay, down

File: test/test-network/systemd-networkd-tests.py (hunk 4 above)
Comment (zhangjian3032) [reply test/test-network/systemd-networkd-tests.py]: okay, May just insert into `test_bond_primary_slave ()` is okay, done

**Comment by zhangjian3032** [comment]
There is a `TEST-29-PORTABLE' CI error here, and I don't know what the reason is.  Could you kindly have a look. @mrc0mmand

**Comment by yuwata** [comment]
> There is a `TEST-29-PORTABLE' CI error here, and I don't know what the reason is. Could you kindly have a look. 

The test is now quite unstable. Please ignore.

**Comment by yuwata** [comment]
CI failures are unrelated. Merging.

**Pull request merged**